// built-extension-sdk/controller/application-user.js
var hookResponseHelper = require('../lib/hook-response-helper')
var endpointsRegex     = require('../endpoints-regex')

module.exports = {
	before: function(req, res){
		req.bclass = getAppUserClass(req.builtApp)
	
		hookResponseHelper.process.call(this, req, res, endpointsRegex.singleAppUserEndpoint, endpointsRegex.multiAppUserEndpoint)
		return utils.Promise.resolve()
	},
	GET: {
		_pre: function(req, res){
			req.bobjekt = req.bclass.Query(req.payload.object)
			return utils.Promise.resolve()
		},
		_post: function(req, res){
			req.bobjekt = req.payload.object
			return utils.Promise.resolve()
		}
	},
	DELETE: {
		_pre: function(req, res){
			req.bobjekt = req.bclass.Query(req.payload.object)
			return utils.Promise.resolve()
		},
		_post: function(req, res){
			req.bobjekt = req.payload.object
			return utils.Promise.resolve()
		}
	},
	POST: {
		_pre: function(req, res){
			req.bobjekt = req.bclass.Object(req.payload.object)
			return utils.Promise.resolve()
		},
		_post: function(req, res){
			req.bobjekt = req.payload.object
			return utils.Promise.resolve()
		}
	},
	'/:userUid': {
		GET:{
			_pre: function(req, res){
				req.bobjekt      = req.bclass.Object(req.params.userUid)
				req.bobjekt.data = req.payload.previousObject
				return utils.Promise.resolve()
			},
			_post: function(req, res){
				req.bobjekt = req.payload.object
				return utils.Promise.resolve()
			}
		},
		PUT: {
			_pre: function(req, res){
				req.bobjekt      = req.bclass.Object(req.params.userUid).assign(req.payload.object)
				req.bobjekt.data = req.payload.previousObject
				return utils.Promise.resolve()
			},
			_post: function(req, res){
				req.bobjekt = req.payload.object
				return utils.Promise.resolve()
			}
		},
		DELETE: {
			_pre: function(req, res){
				req.bobjekt      = req.bclass.Object(req.params.userUid)
				req.bobjekt.data = req.payload.previousObject
				return utils.Promise.resolve()
			},
			_post: function(req, res){
				req.bobjekt = req.payload.object
				return utils.Promise.resolve()
			}
		}
	}
}

function getAppUserClass(bapp) {
  return bapp.Class('built_io_application_user')
}
